python/binding/tritonserver_pybind.cc

```diff
@@ -0,0 +1,1962 @@
+};
+struct AlreadyExists : public TritonError {
+  explicit AlreadyExists(const std::string& what) : TritonError(what) {}
+};
+
+TRITONSERVER_Error*
+CreateTRITONSERVER_ErrorFrom(const py::error_already_set& ex)
+{
+  // Reserved lookup to get Python type of the exceptions,
+  // 'TRITONSERVER_ERROR_UNKNOWN' is the fallback error code.
+  // static auto uk = py::module::import("triton_bindings").attr("Unknown");
+  static auto it = py::module::import("triton_bindings").attr("Internal");
+  static auto nf = py::module::import("triton_bindings").attr("NotFound");
+  static auto ia =
+      py::module::import("triton_bindings").attr("InvalidArgument");
+  static auto ua = py::module::import("triton_bindings").attr("Unavailable");
+  static auto us = py::module::import("triton_bindings").attr("Unsupported");
+  static auto ae = py::module::import("triton_bindings").attr("AlreadyExists");
+  TRITONSERVER_Error_Code code = TRITONSERVER_ERROR_UNKNOWN;
+  if (ex.matches(it.ptr())) {
+    code = TRITONSERVER_ERROR_INTERNAL;
+  } else if (ex.matches(nf.ptr())) {
+    code = TRITONSERVER_ERROR_NOT_FOUND;
+  } else if (ex.matches(ia.ptr())) {
+    code = TRITONSERVER_ERROR_INVALID_ARG;
+  } else if (ex.matches(ua.ptr())) {
+    code = TRITONSERVER_ERROR_UNAVAILABLE;
+  } else if (ex.matches(us.ptr())) {
+    code = TRITONSERVER_ERROR_UNSUPPORTED;
+  } else if (ex.matches(ae.ptr())) {
+    code = TRITONSERVER_ERROR_ALREADY_EXISTS;
+  }
+  return TRITONSERVER_ErrorNew(code, ex.what());
+}
+
+void
+ThrowIfError(TRITONSERVER_Error* err)
+{
+  if (err == nullptr) {
+    return;
+  }
+  std::shared_ptr<TRITONSERVER_Error> managed_err(
+      err, TRITONSERVER_ErrorDelete);
+  std::string msg = TRITONSERVER_ErrorMessage(err);
+  switch (TRITONSERVER_ErrorCode(err)) {
+    case TRITONSERVER_ERROR_INTERNAL:
+      throw Internal(std::move(msg));
+    case TRITONSERVER_ERROR_NOT_FOUND:
+      throw NotFound(std::move(msg));
+    case TRITONSERVER_ERROR_INVALID_ARG:
+      throw InvalidArgument(std::move(msg));
+    case TRITONSERVER_ERROR_UNAVAILABLE:
+      throw Unavailable(std::move(msg));
+    case TRITONSERVER_ERROR_UNSUPPORTED:
+      throw Unsupported(std::move(msg));
+    case TRITONSERVER_ERROR_ALREADY_EXISTS:
+      throw AlreadyExists(std::move(msg));
+    default:
+      throw Unknown(std::move(msg));
+  }
+}
+
+template <typename TritonStruct>
+class PyWrapper {
+ public:
+  explicit PyWrapper(TritonStruct* to, bool o) : triton_object_(to), owned_(o)
+  {
+  }
+  PyWrapper() = default;
+  // Destructor will be defined per specialization for now as a few
+  // Triton object delete functions have different signatures, which
+  // requires a function wrapper to generalize the destructor.
+
+  // Use internally to get the pointer of the underlying Triton object
+  TritonStruct* Ptr() { return triton_object_; }
+
+  DISALLOW_COPY_AND_ASSIGN(PyWrapper);
+
+ protected:
+  TritonStruct* triton_object_{nullptr};
+  bool owned_{false};
```

Comment:
I was wondering whether it is possible to do the same thing with `shared_ptr` and custom deleters. For example, 

```python
class PyParameter {
  ...
  std::shared_ptr<TRITONSERVER_Parameter, TRITONSERVER_ParameterDelete> pointer_;
};
```

We would have to allow COPY_AND_ASSIGN for this to work.

Reply:
It may work for objects that doesn't involve release of ownership, but for request object, user may perform the following and cause `TRITONSERVER_InferenceRequest` being deleted:
```python
req = triton_bindings.TRITONSERVER_InferenceRequest(...) # and initialize
req_2 = req
server.infer_async(req)
del req
del req_2
```
In current implementation of `infer_async`, it informs `req` that it no longer owns the Triton object, but such information is not passed to `req_2` and it will delete the Triton object while it is needed for inference.

Reply:
Actually I think in Python it is just reference counting the same object so `req_2` and `req` should be in sync. Although, just because I am less familiar with binding code in general,I still prefer to keep `DISALLOW_COPY_AND_ASSIGN` to make sure the ownership is controlled within c++ side.